go/pkg/casng/batching.go

```diff
@@ -89,15 +95,195 @@ func (u *BatchingUploader) MissingBlobs(ctx context.Context, digests []digest.Di
 // ctx is used to make and cancel remote calls.
 // This method does not use the uploader's context which means it is safe to call even after that context is cancelled.
 //
-// size is used to toggle compression as well as report some stats. It must reflect the actual number of bytes r has to give.
+// Compression is turned on based the resource name.
+// size is used to report some stats. It must reflect the actual number of bytes r has to give.
 // The server is notified to finalize the resource name and subsequent writes may not succeed.
 // The errors returned are either from the context, ErrGRPC, ErrIO, or ErrCompression. More errors may be wrapped inside.
 // If an error was returned, the returned stats may indicate that all the bytes were sent, but that does not guarantee that the server committed all of them.
 func (u *BatchingUploader) WriteBytes(ctx context.Context, name string, r io.Reader, size int64, offset int64) (Stats, error) {
-	panic("not yet implemented")
+	if !u.streamThrottle.acquire(ctx) {
+		return Stats{}, ctx.Err()
+	}
+	defer u.streamThrottle.release()
+	return u.writeBytes(ctx, name, r, size, offset, true)
 }
 
 // WriteBytesPartial is the same as WriteBytes, but does not notify the server to finalize the resource name.
 func (u *BatchingUploader) WriteBytesPartial(ctx context.Context, name string, r io.Reader, size int64, offset int64) (Stats, error) {
-	panic("not yet implemented")
+	if !u.streamThrottle.acquire(ctx) {
+		return Stats{}, ctx.Err()
+	}
+	defer u.streamThrottle.release()
+	return u.writeBytes(ctx, name, r, size, offset, false)
+}
+
+func (u *uploader) writeBytes(ctx context.Context, name string, r io.Reader, size int64, offset int64, finish bool) (Stats, error) {
```

Comment:
optional: you can shorten a method's signature by writing `size, offset int64`

Reply:
Done